src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2680,21 +2691,68 @@ class ShenandoahUpdateHeapRefsTask : public AbstractGangTask {
     if (_concurrent) {
       ShenandoahConcurrentWorkerSession worker_session(worker_id);
       ShenandoahSuspendibleThreadSetJoiner stsj(ShenandoahSuspendibleWorkers);
-      do_work();
+      do_work(worker_id);
     } else {
       ShenandoahParallelWorkerSession worker_session(worker_id);
-      do_work();
+      do_work(worker_id);
     }
   }
 
 private:
-  void do_work() {
+  void do_work(uint worker_id) {
     ShenandoahHeapRegion* r = _regions->next();
     ShenandoahMarkingContext* const ctx = _heap->complete_marking_context();
+
     while (r != NULL) {
       HeapWord* update_watermark = r->get_update_watermark();
       assert (update_watermark >= r->bottom(), "sanity");
-      if (r->is_active() && !r->is_cset()) {
+
+      // Eventually, scanning of old-gen memory regions for the purpose of updating references can happen
+      // concurrently.  This can be done during concurrent evacuation of roots for example.
+      if (r->is_active() && (r->affiliation() == ShenandoahRegionAffiliation::OLD_GENERATION) && !r->is_cset()) {
+
+        // Note that we use this code even if we are doing an old-gen collection and we have a bitmap to
+        // represent marked objects within the heap region.
+        //
+        // It is necessary to process all objects rather than just the marked objects during update-refs of
+        // an old-gen region as part of an old-gen collection.  Otherwise, a subseqent update-refs scan of
+        // the same region will see stale pointers and crash.
+        //
+        // Kelvin believes (but has not confirmed) the following:
+        //   r->top() represents the upper end of memory that has been allocated within this region.
+        //       As new objects are allocated, the value of r->top() increases to accomodate each new
+        //       object.
+        //   r->get_update_watermark() represents the value that was held in r->top() at the start of
+        //       evacuation.  During evacuation, new objects may be allocated within this heap region
+        //       and this will cause r->top() to increase.  But any objects allocated during the evacuation
+        //       phase do not need to be scanned by update-refs because the to-space invariant is in force
+        //       during evacuation and this will assure that any objects residing between
+        //       r->get_update_watermark() and r->top() hold no pointers to from-space.
```

Comment:
We do need to scan objects that have been allocated by evacuation, because they may hold from-space refs. For this reason, we bump-up UWM when allocating for evac (in ShFreeSet, iirc).